#'Calculate the height of letters in columns
#'
#'Takes the seqDF (see \code{\link{get_seq_df}}) data frame and calculates first relative frequence
#'and then logo height for each amino acid in each column 
#'
#'@param asrStructure the named list returned by \code{\link{runASR}}, \code{\link{loadASR}} or \code{\link{reduce_alphabet}}. Set this to NULL
#'to specify other variables
#'@param seqDF the dataframe generated by \code{\link{get_seq_df}}
#'@param alphabet choice of AA, DNA or RNA
#'@param sequences the sequences to be used when calculating column sequence heights
#'
#'@return a dataframe with three columns representing the height of each AA at each column in the alignment. Formatted to
#' be passed to \code{\link{plot_logo_aln}}\cr
#' heightDF$AA\cr
#' heightDF$Column \cr
#' heightDF$Height \cr
#' #Example\cr
#' heightDF[1, ] = "A", "5", "1.17" - An A in column 5 has a height of 1.17\cr
#' 
#' @examples
#' data(asrStructure)
#' 
#' logo_height_aln(asrStructure)
#' 
#' alphabet <- c("H", "C", "N", "P", "R") #Order is irrelevant
#' logo_height_aln(asrStructure, alphabet=alphabet)
#' 
#' #this method attempts to detect the alphabet but allows the user to specify
#' logo_height_aln(asrStructure, alphabet="DNA")
#' 
#' #retrieve example file stored in the package
#' fasta_file <- system.file("extdata", "runASR_aln_full.fa", package="ASR")
#' #alternatively, specify the filename as a string
#' #fasta_file <- "id_aln_full.fa"
#'
#' # read in the fasta file of choice
#' fasta <- read_fasta(NULL, aln_file = fasta_file) 
#' # convert this to a dataframe that can be read by plot_aln
#' seqDF <- get_seq_df(NULL, fastaDF = fasta) 
#' logo_height_aln(NULL, seqDF = seqDF) #if you want to specify your own file
#'
#' @export

logo_height_aln <- function(asrStructure, seqDF = NULL, alphabet = NULL, alphabetType = NULL, sequences = NULL) {

  shannon <- function(x) {
    x = as.numeric(x)
    x = x * log2(x)
    x
  }
  
  seqDF <- dfError(asrStructure, "seqDF", seqDF, c("Column", "AA", "Label"), "Joint")
  
  if (!is.null(sequences)) {
    seqDF <- seqDF[seqDF$Labels %in% sequences, ]
    seqDF$Column <- factor(seqDF$Labels)
  }
  
  aa = c("A", "C", "D", "E", "F", "G", "H", "I", "K", "L", "M", "N", "P", "Q", "R", "S", "T", "V", "W", "Y")
  dna = c("A", "T", "G", "C")
  rna = c("A", "U", "G", "C")
  if (is.null(alphabetType)) {
    if (is.null(alphabet)) {
      print("No provided alphabetType or alphabet, attempting to detect alphabet")
      default = FALSE
      if ("U" %in% seqDF$AA) {
        alphabetType = "RNA"
        alphabet = rna
        print(paste("Detected alphabetType: ", alphabetType, ". Alphabet equal to: ", seq = ""))
        print(alphabet)
        print("If this is incorrect, please specify an alphabet or alphabet type")
      } else if (Reduce('&',dna %in% seqDF$AA) && length(levels(seqDF$AA)) <= 4) {
        alphabetType = "DNA"
        alphabet = dna
      } else if (Reduce('&',aa %in% seqDF$AA) && length(levels(seqDF$AA)) > 4) {
        alphabetType = "AA"
        alphabet = aa
      } else {
        default = TRUE
        print(paste("Having trouble detecting alphabet in hidden.logo_height_distrib. Using default value for alphabet: AA.
                    If this is incorrect, please specify an alphabet or alphabet type", sep=""))
        alphabet = aa
      }  
      if (!default) {
        print(paste("Detected alphabetType: ", alphabetType, ". Alphabet equal to: ", seq = ""))
        print(alphabet)
        print("If this is incorrect, please specify an alphabet or alphabet type")
      }
    } else {
      alphabetType = "User"
    }
  } else if (alphabetType == "AA") {
    alphabet = aa
  } else if (alphabetType == "DNA") {
    alphabet = dna
  } else if (alphabetType == "RNA") {
    alphabet = rna
  } else if (alphabetType == "User") {
    alphabet = alphabet
  } else {
    stop("Invalid alphabetType for hidden.logo_height")
  }  
  
  #Calculate relative frequency
  res = matrix(rep(0,3),nrow=1,ncol=3)
  for (i in seq(1, length(levels(seqDF$Column)), 1)) {
    col = table(factor(seqDF[seqDF$Column==i, ]$AA))
    freq = col/sum(col)
    for (j in seq(1, dim(freq),1)) {
      res <- rbind(res, matrix(c(names(freq)[j], i, as.vector(freq)[j]),nrow=1,ncol=3))
    }    
  }
  res <- res[-1,]
  res <- data.frame(res[,1], res[,2], as.numeric(res[,3]))
  colnames(res) <- c("AA", "Column", "Probability")
  resProb <- res
  
  res <- res[!is.na(res$Probability), ] ##Remove NAs
  res$Column <- factor(res$Column)
  cols <- levels(res$Column)
  
  fullHeights <- rep(0, length(cols))
  for (i in seq(1, length(cols), 1)) {
    probs <- res[res$Column == cols[i], ]$Probability  
    entropy <- -sum(sapply(probs, shannon)) #calculate Shannon entropy
    infoContent <- log2(length(alphabet)) - entropy #calculate information content
    res[res$Column == cols[i], 3] <- res[res$Column == cols[i], 3] * infoContent
  }
  
  colnames(res) <-  c("AA", "Column", "Height")
  res
}

#'Plot amino acid heights as a barchart to represent logo
#'
#'Takes the sequence height data frame and plot a barchart for each column using the calculated heights
#'
#'@param asrStructure the named list returned by \code{\link{runASR}}, \code{\link{loadASR}} or \code{\link{reduce_alphabet}}. Set this to NULL
#'to specify other variables
#'@param heightDF a data frame containing heights generated by \code{\link{logo_height_aln}}
#'@param colour the colour scheme to be used to colour AAs
#'@param columns a vector containing the column numbers of interest. By default = NULL and all columns are displayed
#'@param sequences the sequences to be used when calculating column sequence heights
#'
#'@return plots a variation on a sequence logo based on seqHeights of AAs in columns
#'
#'@examples
#'data(asrStructure)
#'
#'plot_logo_aln(asrStructure) #if you want to use the information stored in the structure
#'
#' #retrieve example file stored in the package
#' fasta_file <- system.file("extdata", "runASR_aln_full.fa", package="ASR")
#' #alternatively, specify the filename as a string
#' #fasta_file <- "id_aln_full.fa"
#'
#' # read in the fasta file of choice
#' fasta <- read_fasta(NULL, aln_file = fasta_file) 
#' # convert this to a dataframe that can be read by plot_aln
#' seqDF <- get_seq_df(NULL, fastaDF = fasta) 
#'heightDF <- logo_height_aln(NULL, seqDF = seqDF)
#'plot_logo_aln(NULL, heightDF = heightDF)
#'
#'#to specify particular columns to include in figure
#'columns = c(3,4,5,6)
#'
#'plot_logo_aln(asrStructure, columns = columns)
#'#OR
#'plot_logo_aln(NULL, heightDF = heightDF, columns = columns)
#'
#'#to change the colouring of AAs
#'plot_logo_aln(asrStructure, colour = "clustal")
#'plot_logo_aln(asrStructure, colour = "zappo")
#' 
#'@export

plot_logo_aln <- function(asrStructure, heightDF = NULL, colour = "taylor", columns=NULL, sequences=NULL) {
  
  Height <- NULL; rm(Height);
  Column <- NULL; rm(Column);
  AA <- NULL; rm(AA);
  
  heightDF <- dfError(asrStructure, "seqHeights", heightDF, c("Column", "AA", "Height"), "Joint")
  
  if (!is.null(sequences)) {
    #have to recalculate heights
    print("Recalculating height for new sequences, please wait")
    if (is.null(asrStructure)) {
      stop(paste("Cannot calculate heights for reduced sequences without asrStructure. Providing heightDF alone is not adequate."))
    } else {
      heightDF <- logo_height_aln(asrStructure)
    }
  }
  
  if (!is.null(columns)) {
    heightDF <- heightDF[heightDF$Column %in% columns, ]
    heightDF$Column <- factor(heightDF$Column)
  }
  
  heightDF$Column <- factor(heightDF$Column, levels = as.character(sort(as.numeric(levels(heightDF$Column)))))
  heightDF$AA <- factor(heightDF$AA)
  
  if (colour == "clustal") {
    colourPalette = colours_clustal(levels(heightDF$AA))
  } else if (colour == "zappo") {
    colourPalette = colours_zappo(levels(heightDF$AA))
  } else if (colour == "taylor") {
    colourPalette = colours_taylor(levels(heightDF$AA))
  } else if (colour == "mixed") {
    colourPalette = colours_yos(levels(heightDF$AA))
  } else {
    stop("Invalid colour choice for plot_aln()")
  }
  
  ggplot2::ggplot(heightDF, ggplot2::aes(x=Column, y=Height)) + 
    ggplot2::geom_bar(ggplot2::aes(fill = AA), stat = "identity") + 
    ggplot2::scale_fill_manual(values = colourPalette) + 
    ggplot2::theme_bw()+
    ggplot2::theme(text = ggplot2::element_text(size=30),axis.text.x=ggplot2::element_text(angle = 45, hjust=1))
}
