<!DOCTYPE html>
<html lang="en" xmlns:th="http://www.thymeleaf.org">
<body>
<div th:fragment="tutorial" class="panel panel-default" aria-expanded="false" id="tutorial">
    <a id="tutorial-nav"></a>
    <div class="content-section-b" id="tutorial-div">
        <div class="container">
            <div class="row">
                <div class="col-xs-12">

                    <h3 id="feature-log"> New features</h3>
                    <p>GRASP has a dedicated log of updated features - see <a href="/feature_log">Feature log</a> </p>

                    <h3 id="getting-started-nav">Getting started</h3>

                    <p>GRASP infers ancestral proteins from homologous protein sequences. It efficiently determines character states, and the most supported insertions and deletions; GRASP presents all this as partial-order graphs using a visual interface, connected to a phylogenetic tree, with which the user can interact.</p>
                    <h3 id="input-nav">Input data</h3>
                    <p>GRASP requires only two input files and one parameter to run successfully. On GRASP's landing page, the two files you will require are a set of (aligned) sequences and the corresponding phylogenetic tree.</p>
                    <p>GRASP accepts the following file formats.</p>
                    <table class="table">
                        <thead><tr>
                            <th>Sequences</th>
                            <th>Alignment</th>
                            <th>Phylogenetic tree</th>
                        </tr>
                        </thead>
                        <tbody>
                        <tr>
                            <td><a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a></td>
                            <td><a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a>, Clustal</td>
                            <td><a href="https://en.wikipedia.org/wiki/Newick_format">Newick</a></td>
                        </tr>
                        </tbody>
                    </table>
                    <p>To try GRASP out, on the landing page you are presented with the option of running test data (see <a href="#home-img">(1)</a> in <a href="#home-img">Figure 1</a>).
                        This will automatically load an alignment file and tree file; you can then run joint (or marginal) reconstruction allowing you to explore the results provided by GRASP.
                        As we proceed through this tutorial, we will refer to the results of running the test data labelled 'Afriat-Jurnou et al. (29)'.
                        Data sets are taken from publications listed at the bottom of this page</p>
                    <h3 id="help-nav">Help</h3>
                    <p>If, at any stage, you require specific information about what you are seeing presented on the GRASP home or results page,
                        you can hit the help button to switch on help boxes (see <a href="#home-img">(8)</a> in <a href="#home-img">Figure 1</a>).</p>
                    <h3 id="job-name-nav">Job name</h3>
                    <p>To allow you to keep track of your results, on the GRASP landing page you are able to provide a name for your job  (see <a href="#home-img">(2)</a> in <a href="#home-img">Figure 1</a>). This name will be used as an identifier in any saved output data.</p>
                    <h3 id="align-seq-nav">Sequence alignment</h3>
                    <p>Once you have a name, you can 'Select sequence alignment file..' (see <a href="#home-img">(3)</a> in <a href="#home-img">Figure 1</a>). Accepted alignment formats are noted in the table above.</p>
                    <h4>Align using GRASP</h4>
                    <p>If you do not have a set of aligned sequences, GRASP can align your sequences for you. To do this, first switch 'Align extant sequences' to 'Yes' (see <a href="#home-img">(3a)</a> in <a href="#home-img">Figure 1</a>) which will, in turn, allow submission of a sequence library file (see <a href="#home-img">(3b)</a> in <a href="#home-img">Figure 1</a>). Your input library of sequences should be in FASTA format.</p>
                    <p>While you can use GRASP to align your library of sequences, depending on your workflow, you may want to use any of the multiple sequence alignment tools/services listed below::</p>
                    <ul>
                        <li><a href="http://probcons.stanford.edu">ProbCons</a></li>
                        <li><a href="http://tcoffee.crg.cat">T-Coffee</a></li>
                        <li><a href="http://mafft.cbrc.jp/alignment/server/">MAFFT</a></li>
                        <li><a href="http://www.ebi.ac.uk/Tools/msa/muscle/">MUSCLE</a></li>
                        <li><a href="http://dialign.gobics.de">DIALIGN</a></li>
                        <li><a href="http://www.clustal.org/omega/">Clustal Omega</a></li>
                    </ul>
                    <p>We are currently working on a suite of tools that will perform advanced alignment and phylogenetic inference from only a set of homologous sequence; they will be available in GRASP once properly evaluated and benchmarked.</p>
                    <h3 id="phylo-tree-nav">Phylogenetic tree</h3>
                    <p>A phylogenetic tree must be provided for standard reconstruction using 'Select phylogenetic tree file...' (see <a href="#home-img">(4)</a> in <a href="#home-img">Figure 1</a>). Accepted tree formats are noted above.</p>
                    <p>Some of the alignment tools listed above can provide phylogenetic trees. Alternatively, some tools/services/packages for generating trees are listed below:</p>
                    <ul>
                        <li><a href="http://www.microbesonline.org/fasttree/">FastTree</a></li>
                        <li><a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a></li>
                        <li><a href="http://www.megasoftware.net/home">MEGA</a></li>
                        <li><a href="https://sco.h-its.org/exelixis/web/software/raxml/index.html">RAxML</a></li>
                        <li><a href="http://mrbayes.sourceforge.net">MrBayes</a></li>
                    </ul>
                    <h3 id="evol-model-nav">Evolutionary model</h3>
                    <p>GRASP can perform reconstruction based on a variety of evolutionary models which can be selected from the drop down menu 'Select other evolutionary model...' (see <a href="#home-img">(5)</a> in <a href="#home-img">Figure 1</a>). By default, JTT is used.</p>
                    <h3 id="recon-type-nav">Reconstruction type</h3>
                    <p>GRASP can perform both <a href="#technical-info">joint</a> and <a href="#technical-info">marginal</a> reconstructions. We suggest running joint reconstruction as the initial reconstruction (see <a href="#home-img">(6)</a> in <a href="#home-img">Figure 1</a>) as we have made marginal reconstruction available to be run on specific nodes in the results section of GRASP.</p>
                    <h3 id="perform-recon-nav">Perform reconstruction</h3>
                    <p>Once test data has been loaded or user data provided, the reconstruction can be started (see <a href="#home-img">(7)</a> in <a href="#home-img">Figure 1</a>).</p>
                </div>

                <a id="home-img"></a>
                <div class="col-xs-12">
                    <div class="img" id="home-img-div"></div>
                </div>
                <figcaption class="figure-caption">Figure 1. The GRASP landing page with a form for (1) trying an example data set, (2) naming your data set, (3) selecting if sequences need to be aligned, and uploading them,
                    (4) uploading a phylogenetic tree, (5) selecting evolutionary model, (6) opting initial reconstruction: joint or marginal, and (7) starting the analysis. </figcaption>



                <h3 id="errors-nav">Errors</h3>
                <p>The following is a list of the possible error messages that can be generated by GRASP and suggestions as to how to proceed if you encounter them.</p>

                <table class="table table-hover">
                    <thead>
                    <tr>
                        <th>Error message</th>
                        <th>Suggested actions</th>
                    </tr>
                    </thead>
                    <tbody>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Incorrect sequence or alignment format (requires FASTA or Clustal format .aln, .fa or .fasta)</strong></li>
                            </ul></td>

                        <td>Your input files are in the incorrect format. Check that they are valid FASTA or Clustal
                            formats with the correct file extensions. Check that your Clustal file has a correct
                            header (starting with the word “CLUSTAL”). If they are in a different format there are
                            tools available online to convert them to FASTA or Clustal
                            (<a href="https://www.ebi.ac.uk/Tools/sfc/emboss_seqret/">https://www.ebi.ac.uk/Tools/sfc/emboss_seqret/</a>).
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Aligned sequences must have the same length</strong></li>
                            </ul></td>

                        <td>An alignment file by definition should contain sequences with the same length. Check
                            that your alignment file contains aligned sequences and that they are all the same
                            length..
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>There was a problem reading the Newick file.</strong></li>
                            </ul></td>

                        <td>Check your Newick file, specifically it is worth checking that you have a single Newick
                            string that contains the correct number of opening and closing brackets, and that you do
                            not have duplicate identifiers.
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>A distance value in your Newick file couldn’t be parsed as a number.</strong></li>
                            </ul></td>

                        <td>GRASP splits Newick files on the “:” symbol and expects a distance to come immediately
                            after each “:” symbol in the file. Check that the indicated value only contains numbers.
                            If your identifiers themselves contain the “:” symbol they should be replaced (taking
                            care to replace them in the corresponding sequence alignment).
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Extant node names must be unique - &lt;...&gt; is duplicated.</strong></li>
                                <li><strong>Sequence names must be unique - &lt;...&gt; - is duplicated.</strong></li>
                                <li><strong>The sequence names in the provided alignment must all have a match.</strong></li>
                            </ul></td>


                        <td style="vertical-align:middle">Either your files contain sequence names that are
                            duplicated or sequence names do not match the names of nodes in the phylogenetic tree.
                            You will be notified of names that are unique for each file. GRASP will not guess what
                            sequence that should be placed in the tree, so make sure the names agree in the two
                            files. Some alignment programs willnamend the names, which means you may need to edit
                            them manually.
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Sequence &lt;...&gt; is using an invalid symbol. Unrecognised symbol: &lt;...&gt; at index &lt;...&gt;</strong></li>
                            </ul></td>

                        <td>Check your sequence alignment and phylogenetic tree files and replace or remove the
                            invalid symbols. In cases where ‘X’ has been used in place of an unknown amino acid
                            these can either be corrected by replacing the ‘X’ symbols with a ‘-‘ symbol or by
                            identifying higher quality sequences without unknown amino acids and realigning and
                            inferring new phylogenetic trees.
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Your files are too large for the web server.</strong></li>
                            </ul></td>

                        <td>Contact us to download and install a local, commandline tool that you can run on your
                            own hardware. This will not have the graphical user interface, however.
                        </td>
                    </tr>
                    </tbody>
                </table>

                <p>You should also be aware that the server only keeps the results for a limited time. You may notice that if you leave the results page open for a long time without interaction, it may stop responding.</p>



                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell"><div class="text_cell_render border-box-sizing rendered_html"><hr></hr></div></div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h2 id="results-nav">GRASP Results</h2>
                            <p>The figures included in this section were generated by running the test data labelled 'Afriat-Jurnou et al. (29)' available on GRASP (see <a href="#home-img">(1)</a> in <a href="#home-img">Figure 1</a>).</p>
                            <h3 id="results-phylo-nav">Phylogenetic tree</h3>
                            <p><a id='phylogenetic_tree'></a></p>
                            <p>The results page shows an interactive version of the phylogenetic tree that was provided for reconstruction. GRASP will number internal phylogenetic node labels from the root node (N0, N1, ...) using the principle of "depth first". By default, extant labels are displayed and the user is given the option to also display or hide node labels and branch lengths (see <a href="#phylo-tree-img">(1)</a>, <a href="#phylo-tree-img">(2)</a> &amp; <a href="#phylo-tree-img">(3)</a> in <a href="#phylo-tree-img">Figure 2</a>). The user can also switch between an additive tree type or a cladogram (see <a href="#phylo-tree-img">(4)</a> in <a href="#phylo-tree-img">Figure 2</a>). This tree can also be saved as a PNG image (see <a href="#phylo-tree-img">(5)</a> in <a href="#phylo-tree-img">Figure 2</a>).</p>
                            <p>Each node is coloured by evolutionary distance according to the legend on the left hand side (see <a href="#phylo-tree-img">(6)</a> in <a href="#home-img">Figure 1</a>). One node will always appear pink and be noted in the header above the tree. This is the node that is having its reconstruction displayed in the graph below the phylogenetic tree and is known as the <strong>primary node</strong>. By default, node N0 (the root node) is the <strong>primary node</strong> and will appear pink and be listed in the header.</p>
                            <a id="phylo-tree-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="phylo-tree-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 2. The phylogenetic tree with options to
                                (1) toggle names on extant nodes,
                                (2) toggle names on internal/ancestral nodes,
                                (3) toggle distances on branches,
                                (4) toggle mode of tree: additive or cladogram, and
                                (5) save a PNG image of the displayed tree. </figcaption>
                            <p>By hovering over nodes in the phylogenetic tree, the name will be revealed. By right clicking an ancestral node in the tree you will bring up a number of
                                operations which can be performed on the ancestor. Note that some operations may take time to complete.</p>
                            <table class="table">
                                <thead>
                                <tr>
                                    <th>Option</th>
                                    <th>Operation</th>
                                </tr>
                                </thead>
                                <tbody>
                                <tr>
                                    <td>View marginal reconstruction</td>
                                    <td>Change graph visualisation to display the marginal reconstruction of this phylogenetic node.
                                        I.e. make this ancestor the <strong>primary, phylogenetic node</strong>.
                                        The node in the phylogenetic tree will become pink and the page header will reflect this selection.</td>
                                </tr>
                                <tr>
                                    <td>View joint reconstruction</td>
                                    <td>Change graph visualisation to display the joint reconstruction of this phylogenetic node. I.e. make this node the <strong>primary, phylogenetic node</strong>.
                                        The node will be drawn pink and the page header will reflect this selection.</td>
                                </tr>
                                <tr>
                                    <td>Add joint reconstruction</td>
                                    <td>Add a graph of another ancestor to the visualisation, based on a current joint reconstruction. Additional graphs are "stacked" and shown with a pink outline. Note that if your primary node is a marginal reconstruction, the newly added reconstruction will become your new primary, joint reconstruction.</td>
                                </tr>
                                <tr>
                                    <td>Collapse subtree</td>
                                    <td>Hide all descendants/children of this phylogenetic node. Use this when your tree is large, and hard to read.</td>
                                </tr>
                                <tr>
                                    <td>Expand subtree</td>
                                    <td>Expand all descendants/children of this phylogenetic node.</td>
                                </tr>
                                </tbody>
                            </table>
                            <div class="col-xs-8 col-xs-offset-2">
                                <div class="col-xs-12">
                                    <div class="img" id="phylo-tree-right-click-img-div"></div>
                                </div>
                                <figcaption class="figure-caption">Figure 3. The pop-up menu of the phylogenetic tree is activated by right-clicking on an internal node. </figcaption>
                            </div>
                            <div class="clearfix"></div>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="motif-searching-nav">Motif Searching</h3>
                            <p> You are able to search for motifs in saved joint reconstructions. </p>

                            <p><h4>Example searches:</h4>
                            <ul>
                                <li>
                                <code> %A%</code>: Highlight nodes with an <code> A </code> in any position, you can also use just an <code> A </code> however the <code>%</code> signs are recommended.
                                </li>
                            <li>
                                <code> %A-*M%</code>: Highlight nodes with an <code> A </code> followed by any number of gaps and then an <code> M </code>.
                            </li>
                            <li>
                                <code> %AM(P|R)%</code>:  Highlight nodes with an <code> A </code> followed by an <code> M </code>, followed by either a <code> P </code> or an <code> R </code>.
                            </li>

                            <li>
                                <code> %A_M%</code>:  Highlight nodes with an <code> A </code> followed by any other character (e.g. a gap, <code> K </code>, <code> P </code> etc) followed by an <code> M </code>.
                            </li>
                            </ul>
                            </p>
                            <p><h4>Syntax guide:</h4>
                            <ul>
                                <li>
                                    <code> _ </code>: Wildcard character.
                                </li>
                                <li>
                                    <code> | </code>:  denotes one of the two options.
                                </li>
                                <li>
                                    <code> *</code>: repetition of preceding amino acid 0 or more times.
                                </li>

                                <li>
                                    <code> {n}</code>: repetition of the preceding amino acid <code>n</code> times.
                                </li>

                                <li>
                                    <code> {m, n}</code>: repetition of the preceding amino acid more than <code>m</code> times and less than <code>n</code> times.
                                </li>
                            </ul>
                            </p>
                            <p>
                                <strong>Note 1:</strong> Since the gap patterns from the original alignment are preserved, unless you specify gaps or use (<code>-*</code>) between your search elements, only those sequences without gaps between the search terms will be highlighted.
                            </p>
                            <p>
                                <strong>Note 2:</strong> The searching looks for a match <em>anywhere</em> in the sequence, this means that even though two sequences are highlighted, they might not have the motif in the same position! So it is recommended you confirm your findings by downloading the joint reconstructions and checking the alignment.
                            </p>

                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="taxonomy-annotation-nav">Taxonomy annotation</h3>
                            <p> GRASP can automatically retrieve taxonomic information from the NCBI and UniProt databases and append this information to the nodes of the phylogenetic tree. Nodes in the tree are assigned the highest shared taxonomic rank of the sequences beneath them. Once nodes
                                are annotated by clicking on the <strong>Annotate taxonomy</strong> button, the highest shared rank will be visible at collapsed nodes. Clicking on a node will report the highest shared taxonomic rank as well as presenting a histogram of the distribution within the
                                taxonomic rank immediately below the highest shared rank</p>

                            <p>For example, in Figure 4, the four sequences contained under the ancestral node N1 all belong the the Phylum Ascomycota, but differ in their taxonomic Class. </p>

                            <div class="col-xs-8 col-xs-offset-2">
                                <div class="col-xs-12">
                                    <div class="img" id="taxonomic-annotation-img-div"></div>
                                </div>
                                <figcaption class="figure-caption">Figure 4. The summary available after annotating nodes with their taxonomic rank. </figcaption>
                            </div>
                            <div class="clearfix"></div>


                            <p>In order to annotate taxonomy from the NCBI and UniProt databases, GRASP expects the sequence IDs in your FASTA file to be in either the NCBI or UniProt ID format -</p>

                            <p><strong>>PRD21445.1 Cyp2u1 [Nephila clavipes]</strong></p>

                            <p>or</p>

                            <p><strong>>tr|A0A2P6K539|A0A2P6K539_NEPCL Cyp2u1 OS=Nephila clavipes OX=6915 GN=NCL1_51699 PE=4 SV=1</strong></p>

                            <p>By default, GRASP will expect the ID to be separated by a whitespace from any additional information.</p>

                            <p>However, it will accept IDs which are completely stripped of additional information - </p>

                            <p><strong>>PRD21445.1</strong></p>

                            <p>or</p>

                            <p><strong>>tr|A0A2P6K539|A0A2P6K539_NEPCL</strong></p>

                            <p>Or IDs that have all whitespace removed (in order to force annotations to appear on phylogenetic trees) <em>if</em> the additional information is separated from the ID by a pipe symbol (|) -</p>

                            <p><strong>>PRD21445.1|Cyp2u1_Nephila_clavipes</strong></p>

                            <p>or</p>

                            <p><strong>>tr|A0A2P6K539|A0A2P6K539_NEPCL|Cyp2u1_OS=Nephila_clavipes</strong></p>


                            <p>Note that this means that headers where additional information is not separated by either a whitespace or a pipe symbol will not be able to have their taxonomy annotated - i.e. the following will <em>not</em> work -</p>


                            <p><strong>>PRD21445.1_Cyp2u1_Nephila_clavipes</strong></p>

                            <p>or</p>

                            <p><strong>>tr|A0A2P6K539|A0A2P6K539_NEPCL_Cyp2u1_OS=Nephila_clavipes</strong></p>


                            <p>All of the existing constraints about sequence names matching between alignment and phylogenetic tree must still be adhered to.</p>



                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="po-graph-nav">Partial-order graph</h3>
                            <p>GRASP uses partial-order graphs to display information about ancestors. They are suited to this task as (a) sequence homology amongst members in a protein family (and by extension their ancestors) may not always map unambiguously to a single, linear ordering, and (b) unresolved insertions and deletions can be identified as alternative paths through the graph.</p>
                            <p>We suggest that you read a partial-order graph is from left-to-right, node-by-node, forming a putative biological sequence (N- to C-terminus) by following a single path. While GRASP attempts to resolve the history of insertions and deletions, at a single node, you may be presented with multiple in- and out-going edges, each of which joins potential ancestral sequences together.</p>

                            <h4>Nodes in a partial-order graph</h4>
                            <p>A node simply represents the character state of a position in a sequence. The position is not defined in an absolute sense, rather in a relative sense when there is evidence of an order. For practical reasons, there is still a positional index to assist navigation in GRASP.</p>
                            <p>The node displays either a single character state or a distribution over possible character states. The states of ancestral nodes are inferred by <a href="/ml">maximum likelihood</a>.</p>

                            <h4>Edges in a partial-order graph and consensus paths</h4>
                            <p><a id='edge_cons_paths'></a></p>
                            <p>An edge represents a valid path to create a full-length sequence. Ultimately, once homologous sections have been determined by alignment,
                                the set of extant sequences express the space of possible paths that can be taken. Of them, by enumerating them at each node,
                                GRASP infers the maximally <em>parsimonious</em> edges, i.e. the edge or edges that imply the least number of changes across the whole evolutionary history.
                                We call this process <em>bi-directional edge-parsimony</em> (BDEP). The term "bi-directional" here refers to the separation of in- and out-going edges of a node,
                                and therefore an edge can be maximally parsimonious, for one or both nodes it joins. You can find a more detailed explanation <a href="#technical-info">here</a>.</p>
                            <p>GRASP identifies edges for which there is qualified support, and makes a number of distinctions that are important to interpret sequences that you assemble through them:</p>
                            <ul>
                                <li>
                                    <p>Dark-coloured edges are fully supported by bi-directional edge-parsimony; these links are strong candidates for how an ancestor can be assembled.
                                        They are not exclusive; multiple paths connecting different parts of a full-length sequence may be present.
                                        Nor are they guaranteed to identify even a single full-length sequence.<br></br></p>
                                </li>
                                <li>
                                    <p>A light-coloured edge is maximally parsimonious when only one of the nodes it joins is considered.
                                        These edges are weaker candidates but are kept to allow full-length ancestors to be formulated in the absence of the stronger option above.</p>
                                </li>
                                <div class="col-xs-12">
                                    <div class="img" id="dark-light-edges-img-div"></div>
                                </div>
                                <div class="clearfix"></div>
                            </ul>
                            <p>GRASP also has access to the raw number of extant sequences on which the inference is based. This can be used to separate edges further:</p>
                            <ul>
                                <li>
                                    <p>Dashed edges have only a single representative in the set of extant sequences; these should almost always be disregarded and are often the result of out-groups that have been used to build the phylogenetic tree.</p>
                                </li>
                                <li>
                                    <p>GRASP provides the option of displaying one <em>consensus</em> sequence. If this option is turned on,
                                        thick/bold edges represent a single, consensus traversal of the graph, giving (greedy) preference to
                                        (a) bi-directional edge-parsimony support (edge is drawn dark) and (b) greater number of extant sequences following, irrespective of their phylogenetic relationship.
                                        (All alternative paths are drawn with a thin outline.)</p>
                                </li>
                            </ul>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="vis-joint-nav">Graph visualisation - joint reconstruction</h3>
                            <p>After running reconstruction, you will be able to see a navigation bar (see <a href="#graph-view-img">(1)</a> in <a href="#graph-view-img">Figure 5</a>), a multiple sequence alignment graph (see <a href="#graph-view-img">(2)</a> in <a href="#graph-view-img">Figure 5</a>) and the primary node reconstruction (see <a href="#graph-view-img">(3)</a> in <a href="#graph-view-img">Figure 5</a>) underneath the phylogenetic tree.</p>
                            <a id="graph-view-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="graph-view-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 5. The reconstruction results panel has multiple sub-panels to view and interact with results, including
                                (1) a bar for viewing the overall variation (optional edges and insertion/deletions) found in the set of reconstructions,
                                (2) a multiple sequence alignment graph,
                                (3) a partial-order graph for the phylogenetic node selected as primary,
                                (4) slider for selecting a section, and zooming in and out of the set of ancestors currently displayed, and buttons to
                                (5) save the graphs graphically,
                                (6) select colours for amino acids as they appear in graphs,
                                (7) perform mutant library analysis (applies to marginal reconstructions only), and
                                (8) reset.
                            </figcaption>
                            <h4>Navigation bar</h4>
                            <p>The navigation bar has a blue box (see <a href="#graph-view-img">(4)</a> in <a href="#graph-view-img">Figure 5</a>) which can be expanded, or collapsed and moved across the window to view different sections of the graphs. The region of the graph that you are viewing is within the blue box. Depending on which node is being viewed as the <strong>primary node</strong> (see <a href="#phylogenetic_tree">Phylogenetic tree</a>) you will see grey boxes and red circles across the navigation bar.</p>
                            <p><strong>Grey boxes</strong> represent variations in insertions and deletions of the ancestor at the primary node that can not be resolved definitively; they indicate the presence of two or more edges in the ancestor at the corresponding position, all supported by maximum parsimony.</p>
                            <p><strong>Red circles</strong> represent inferred deletions in the reconstructed sequence of the ancestor at the primary node, relative to the multiple sequence alignment graph; they represent positions where supported edges determine that deletions of one or more nodes have occurred.</p>
                            <h4>Multiple sequence alignment</h4>
                            <p>The multiple sequence alignment is represented here as a partial-order alignment graph which captures variation across all sequences. Each node in the graph is represented as a pie chart showing the distribution of amino acids at that position across all extant sequences. The most probable amino acid is shown using its one-letter code. When you hover your mouse over each node, a bar chart will display, showing the distribution in more detail, as seen in <a href="#msa-pie-chart-img">Figure 6</a>.</p>
                            <a id="msa-pie-chart-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="msa-pie-chart-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 6. The multiple sequence alignment graph shows the sequence composition of <i>all</i> extant sequences, and what insertion/deletions are implied, in terms of alternative paths over edges.</figcaption>
                            <p>The edges in the multiple sequence alignment graph represent the insertion and deletions of <em>all</em> extant sequences.</p>
                            <h4>Primary node reconstruction</h4>
                            <p>This graph displays the reconstructed partial-order graph of the primary node. More technical information of joint reconstruction is provided <a href="#technical-info">below</a>.</p>
                            <p>Edges in reconstructions are inferred by edge-parsimony. One or more ancestral sequences can be formed by following one path. We recommend that dark edges are given preference, but all presented edges have some support. We recommend that dashed edges are not used as their support is based only on a single sequence (usually representing an out-group). As a last guide, but only if you really want a single sequence, use bold edges as they indicate the path which has the greatest number of extant sequences supporting.</p>
                            <h4>The joint reconstruction stack and the merged reconstruction</h4>
                            <p>When the 'Add joint reconstruction' option is selected for a node in the phylogenetic tree, its reconstructed sequence will be added below the sequence of the primary node (boxed with a blue shade that indicates the evolutionary distance of the phylogenetic node it is taken from to the root). We refer to this as the "stack" and you can keep adding joint reconstructions to it; this allows for comparisons between ancestral states at different phylogenetic nodes.</p>
                            <p>On addition to the aligned visualisation of reconstructions in the stack, a graphical representation of all the ancestral graphs in a "merged" format will appear at the bottom of the graph visualisation as seen in <a href="#merged-graph-img">Figure 7</a>. The character state of each individual graph is displayed as a slice in a pie graph of all the ancestral graphs. Note that the colour of the outline indicates the evolutionary distance to the root, just as the background of the panel where the graph is shown.</p>
                            <a id="merged-graph-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="merged-graph-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 7. After joint reconstruction, all extants, the primary node and all added ancestral nodes are shown as an aligned stack of graphs. The merged graph summarises the make-up of all displayed reconstructions. </figcaption>
                            <h4>Graph options</h4>
                            <p>Graphs can be saved as PNG images using the 'Save graphs' button (see <a href="#graph-view-img">(5)</a> in <a href="#graph-view-img">Figure 5</a>).</p>
                            <p>The amino acid colour scheme is, by default, set to Clustal (see <a href="#graph-view-img">(6)</a> in <a href="#graph-view-img">Figure 5</a>).</p>
                            <p>To view mutants (see <a href="#graph-view-img">(7)</a> in <a href="#graph-view-img">Figure 5</a>), you must first perform a marginal reconstruction on a node (see <a href="#marg_recon">Marginal reconstruction</a>).</p>
                            <p>The graph visualisation can be reset so only the reconstruction of the primary node is displayed (see <a href="#graph-view-img">(8)</a> in <a href="#graph-view-img">Figure 5</a>).</p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="vis-marg-nav">Graph visualisation - marginal reconstruction</h3>
                            <p>A marginal reconstruction can be performed on the root node from the loading page or any node selected in the
                                phylogenetic tree on the results page.
                                A marginal reconstruction will look similar to <a href="#graph-view-marginal-img">Figure 8</a>.
                            </p>
                            <p>The reconstruction of the primary node will be composed of pie charts similar to the MSA plot.
                                This represents the distribution of amino acids at each position in the reconstructed sequence.
                                More technical information about marginal reconstruction is provided <a href="#technical-info">here</a>.
                            </p>
                            <a id="graph-view-marginal-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="graph-view-marginal-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 8. After marginal reconstruction, all extants, and the primary node are shown as graphs.
                                The graph for the primary node reflects the probability of each amino acid at each position.
                                The ability to determine and view alternative mutants around an ancestor is enabled.
                            </figcaption>
                            <h4>Mutants</h4>
                            <p>Once a marginal reconstruction has been performed, it is possible to identify targets for mutational analysis. This is based on the idea that the reconstruction represents just one possible ancestor, and that others can be explored around it by mutating positions that are inferred to be in an ambivalent state.</p>
                            <p>The user can select the level of complexity of the mutant library.
                                The base case "0" means that all nodes map to the single, most probable amino acid.
                                You will note by setting it to "0" all nodes become single-state.
                                By increasing the level, alternative amino acids are added to positions that are more ambivalent;
                                at higher levels of complexity, some positions are assigned two or more mutational targets.
                                An arrow appears close to the location of the identified nodes.
                                You will note that the ambivalent nodes are shown with more character states.
                                More technical information is provided <a href="#technical-info">below</a>.
                            </p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="download-nav">Downloading results</h3>
                            <p>You can download the results generated by GRASP, and input files, through the download pane (as shown below). The download pane can be opened by clicking the "Download Results" drop down button located below the graph visualisation area (see <a href="#download-pane-img">(1)</a> in <a href="#download-pane-img">Figure 9</a>).</p>
                            <a id="download-pane-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="download-pane-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 9. The download results section, with a form to (1) show download specific file options, and what those files may be including (2) the annotated phylogenetic tree, (3) reconstructions, and (4) probability distributions and consensus sequences.</figcaption>
                            <p>To download results, tick the boxes for each file you wish to download and then click the download button at the bottom of the pane.</p>
                            <p>On the left hand side of the download pane ('<strong>Phylogenetic Tree</strong>', see <a href="#download-pane-img">(2)</a> in <a href="#download-pane-img">Figure 9</a>) there is a single option to download the phylogenetic tree. The corresponding downloaded file will simply be the tree you provided but with internal/ancestral nodes labelled (N0, N1..etc.).</p>
                            <p>In the middle section ('<strong>Ancestral Graphs</strong>', see <a href="#download-pane-img">(3)</a> in <a href="#download-pane-img">Figure 9</a>) you will find options to download partial-order graph representations of the following:</p>
                            <ul>
                                <li>Full multiple sequence alignment.</li>
                                <li>Marginal reconstruction for the currently selected primary node.</li>
                                <li>Joint reconstruction of all ancestor sequences. <strong>Warning</strong> - This can take some time to calculate!</li>
                                <li>Joint reconstruction for the currently selected primary node.</li>
                            </ul>
                            <p>Since GRASP represents sequences as partial-order graphs, these files are in the <a href="https://en.wikipedia.org/wiki/DOT_(graph_description_language">'.dot'</a> format. There are a number of programs that you can use to visualise '.dot' files, one of the more popular cross-platform options is <a href="http://www.graphviz.org">'Graphviz'</a>.</p>
                            <p>Finally, on the right hand side of the download pane ('<strong>Sequence Information</strong>', see <a href="#download-pane-img">(4)</a> in <a href="#download-pane-img">Figure 9</a>) you will find options to download information for the following:</p>
                            <ul>
                                <li>distibution of characters in the marginal reconstruction for the primary node. This is in a 'tab separated value' (.tsv) format;</li>
                                <li><a href="#edge_cons_paths">consensus sequence</a> for the marginal reconstruction of the currently selected primary node (.fasta format);</li>
                                <li><a href="#edge_cons_paths">consensus sequence</a> for the joint reconstruction of the currently selected primary node (.fasta format);</li>
                                <li><a href="#edge_cons_paths">consensus sequence</a> for all joint reconstructions (.fasta format);</li>
                                <li>distribution of characters in the multiple sequence alignment provided by the user or generated by GRASP (.tsv format); and</li>
                                <li>multiple sequence alignment.</li>
                            </ul>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h2 id="tech-info-nav">Technical information<a id="technical-info"></a></h2>
                            <h3 id="joint-marg-nav">What is the difference between joint and marginal reconstruction?</h3>
                            <p>GRASP employs exact probabilistic inference; it distinguishes between two types of reconstructions:</p>
                            <p><strong>Joint</strong> - GRASP infers the most probable combination of all ancestral states.
                                The assignment of all ancestral states (across the whole phylogenetic tree) optimises a single objective
                                (the joint state with the greatest <em><a href="/ml">likelihood</a></em>).
                                This will tell you which are the most probable ancestors, given the extant sequences and their phylogenetic relationships.
                                The inference is valid for the whole tree, and it is practical to compare ancestors with one another as they represent
                                snapshots from the same evolutionary history.</p>
                            <p><strong>Marginal</strong> - GRASP infers the probability distribution of each residue for a single, nominated ancestor.
                                This inference will present you with the probability of amino acids at each position in the nominated ancestor,
                                given extant sequences and their phylogenetic relationships. The inference "marginalises" all other ancestors,
                                which means that (a) each ancestor needs to be inferred separately, and (b) resulting distributions are not directly comparable.
                                Multiple marginal reconstructions (for different ancestors) are based on the <em>same</em> input data, but their mutual consistency is not enforced;
                                they allow for <em>different</em> versions of evolutionary histories along the <em>same</em> branches.</p>
                            <p>Once GRASP performs a joint reconstruction it will determine and keep all ancestral states in memory; this process can take some time to complete.
                                "Adding" nodes on the results page will not require more inference of ancestral states, so the delay you experience is minor.
                                Marginal reconstruction needs to be performed for each node; inference is separate so there will be longer delays each time
                                even after you have moved to the results page.</p>

                            <h3 id="bi-pars-nav">What is bi-directional edge-parsimony?</h3>
                            <p>Edges represent optional paths through which sequences can be composed;
                                ultimately, the collective of extant sequences when aligned determines the set of all possible paths;
                                that set is the source from which we draw the path or paths for any ancestor.
                                We are interested in identifying which edges that are best supported for an ancestor, given their phylogenetic relationships.
                                For this we use the basic principle of maximum parsimony, where only the most parsimonious edges are kept for each individual ancestor.</p>
                            <p>Maximum parsimony is applied to edges of a node; for each node, we separately process edges labelled by the nodes preceding it, and edges labelled by the nodes following it.
                                An edge can therefore be supported in one direction but not the other.
                                Uni-directional support is sufficient for GRASP to display an edge and maintain any paths to the nodes concerned;
                                bi-directional support should be seen as a much stronger endorsement though.
                                We run maximum parsimony to identify <em>all</em> edges that achieve the best score (the same score can be shared by multiple edge selections).
                                This means that in many cases multiple paths through an ancestor cannot be ruled out.</p>

                            <h3 id="mutant-nav">How are mutant libraries determined?</h3>
                            <p>A marginal reconstruction allows the user to identify the positions that should be mutated with priority, to explore the space of probable ancestors.</p>
                            <p>Initially, the complexity of the mutant library is "0", suggesting for each position, only the most probable amino acid.
                                At "1", a single mutation is allowed, at the position that best improves the summed <a href="https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence">Kullback-Leibler divergence</a> between the
                                mutant distributions and the corresponding marginal distributions, across all sequence positions.
                                For each increase in the level, another mutation is identified where an alternative amino acid is suggested.
                                Note that additional mutations could occur at a position that has been mutated previously or in a different position.
                                Either way, the potential library will increase in complexity, i.e. increase the number of theoretically possible mutants.
                            </p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h2 id="advanced-nav">Advanced Concepts</h2>
                            <a href="/ml">What is maximum likelihood?</a>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h2 id="refs-nav">References</h2>
                            <ol>
                                <li>Afriat-Jurnou L, Jackson CJ and Tawfik DS. Reconstructing a Missing Link in the Evolution of a Recently Diverged Phosphotriesterase by Active-Site Loop Remodeling.
                                    <em>Biochemistry</em> 2012, 51, 6047-6055. <a href="http://dx.doi.org/10.1021/bi300694t">DOI:10.1021/bi300694t</a></li>
                                <li>Clifton BE and Jackson CJ. Ancestral protein reconstruction yields insights into adaptive evolution of binding specificity in solute-binding proteins.
                                    <em>Cell Chem Biol</em>, 23(2): 236-245. <a href="http://dx.doi.org/10.1016/j.chembiol.2015.12.010">DOI:10.1016/j.chembiol.2015.12.010</a></li>
                                <li>Hudson WH, Kossmann BR, de Vera IM, Chuo SW, Weikum ER, Eick GN, Thornton JW, Ivanov IN, Kojetin DJ, and Ortlund EA.
                                    Distal substitutions drive divergent DNA specificity among paralogous transcription factors through subdivision of conformational space.
                                    <em>Proc Natl Acad Sci USA</em>, 113(2): 326-331. <a href="http://dx.doi.org/10.1073/pnas.1518960113">DOI:10.1073/pnas.1518960113</a></li>
                            </ol>
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <!-- /.container -->
    </div>
</div>
</body>
</html>