<!DOCTYPE html>
<html lang="en" xmlns:th="http://www.thymeleaf.org">
<body>
<div th:fragment="tutorial" class="panel panel-default" aria-expanded="false" id="tutorial">
    <a id="tutorial-nav"></a>
    <div class="content-section-b" id="tutorial-div">
        <div class="container">
            <div class="row">
                <div class="col-xs-12">

                    <h3 id="feature-log">GRASP is alive</h3>
                    <p>GRASP has a web site for code, associated tools, data sets, and support information - see <a href="https://bodenlab.github.io/GRASP-suite">GRASP-suite</a>.
                        This site is updated to reflect changes to the service and resources around it.
                        You can peruse and submit issues for the web service <a href="https://github.com/bodenlab/GRASP/issues">GRASP/issues</a> and the command-line tool <a href="https://github.com/bodenlab/bnkit/issues">bnkit/issues</a>.
                        If you use GRASP in your work please cite the paper listed <a href="#refs-grasp">below</a>.
                    </p>

                    <h3 id="getting-started-nav">Getting started</h3>

                    <p>GRASP infers ancestral proteins from homologous protein sequences--a process known as ancestral sequence reconstruction.
                        (It does <em>not</em> work with DNA at the moment, though we expect to introduce this in the future.)
                        GRASP efficiently determines ancestral character states, and the most supported insertions and deletions.
                        GRASP presents all this as partial-order graphs using a visual interface, connected to a phylogenetic tree, with which the user can interact.</p>
                    <p>Many features are only available as a result of <em>saving</em> an ancestor reconstruction. To enable this, you need to be signed in as a user.
                        We strongly recommend that you do this as the workflow will then be un-interrupted.
                        Use the 'Login' menu to start the registration process.
                        Note that the web service will log you out automatically if you are idle for a longer period of time.
                    In that case, login again, and your saved reconstructions will be available from the menu item marked with your user name.
                    </p>
                    <p>Passwords are stored in an encrypted form, but our strong advise is to use a unique password as the http protocol is subject to some security risks.
                    </p>
                    <h3 id="input-nav">Input data</h3>
                    <p>GRASP requires only two input files and one parameter to run successfully.
                        On GRASP's landing page, you will require two files: a set of (aligned) sequences and the corresponding phylogenetic tree.</p>
                    <p>GRASP accepts the following file formats.</p>
                    <table class="table">
                        <thead><tr>
                            <th>Alignment</th>
                            <th>Phylogenetic tree</th>
                        </tr>
                        </thead>
                        <tbody>
                        <tr>
                            <td><a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a>, Clustal</td>
                            <td><a href="https://en.wikipedia.org/wiki/Newick_format">Newick</a></td>
                        </tr>
                        </tbody>
                    </table>
                    <p>To try GRASP out, on the landing page you are presented with the option of using example data (see <a href="#home-img">(1)</a> in <a href="#home-img">Figure 1</a>).
                        This will automatically load an alignment file and tree file; you can then run the reconstruction allowing you to explore the results provided by GRASP.
                        We will use results from running the test data labelled 'Afriat-Jurnou et al. (29)' and 'CYP2U1 (595)' for the figures below.
                        Data sets are taken from publications listed at the bottom of this page</p>
                    <h3 id="help-nav">Help</h3>
                    <p>If, at any stage, you require 'on-screen tips' about what you are seeing presented on the GRASP home or results page,
                        you can hit the help button to switch on help texts (see <a href="#home-img">(8)</a> in <a href="#home-img">Figure 1</a>).</p>
                    <h3 id="job-name-nav">Job name</h3>
                    <p>To allow you to keep track of your results, on the GRASP landing page you must provide a unique name for your job (see <a href="#home-img">(2)</a> in <a href="#home-img">Figure 1</a>). This name will be used as an identifier in any saved output data.</p>
                    <h3 id="align-seq-nav">Sequence alignment</h3>
                    <p>Once you have a name, you can 'Select sequence alignment file...' (see <a href="#home-img">(3)</a> in <a href="#home-img">Figure 1</a>). Accepted alignment formats are noted in the table above.</p>
                    <p>You may want to use any of the multiple sequence alignment tools/services in the non-exhaustive list below to create your alignment file.</p>
                    <ul>
                        <li><a href="http://probcons.stanford.edu">ProbCons</a></li>
                        <li><a href="http://tcoffee.crg.cat">T-Coffee</a></li>
                        <li><a href="http://mafft.cbrc.jp/alignment/server/">MAFFT</a></li>
                        <li><a href="http://www.ebi.ac.uk/Tools/msa/muscle/">MUSCLE</a></li>
                        <li><a href="http://dialign.gobics.de">DIALIGN</a></li>
                        <li><a href="http://www.clustal.org/omega/">Clustal Omega</a></li>
                    </ul>
                    <h3 id="phylo-tree-nav">Phylogenetic tree</h3>
                    <p>A phylogenetic tree must be provided using 'Select phylogenetic tree file...' (see <a href="#home-img">(4)</a> in <a href="#home-img">Figure 1</a>). Accepted tree formats are noted above.</p>
                    <p>Some of the alignment tools listed above can provide phylogenetic trees. Alternatively, some tools/services/packages for creating trees from your alignment are listed below.</p>
                    <ul>
                        <li><a href="http://www.microbesonline.org/fasttree/">FastTree</a></li>
                        <li><a href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</a></li>
                        <li><a href="http://www.megasoftware.net/home">MEGA</a></li>
                        <li><a href="https://sco.h-its.org/exelixis/web/software/raxml/index.html">RAxML</a></li>
                        <li><a href="http://mrbayes.sourceforge.net">MrBayes</a></li>
                    </ul>
                    <h3 id="evol-model-nav">Evolutionary model</h3>
                    <p>GRASP can perform reconstruction based on a variety of evolutionary models that can be selected from the drop down menu 'Select other evolutionary model...'
                        (see <a href="#home-img">(5)</a> in <a href="#home-img">Figure 1</a>). By default, JTT (Jones-Taylor-Thornton) is used, which is commonly used for the analysis of amino acid sequences.</p>
                    <h3 id="recon-type-nav">Reconstruction type</h3>
                    <p>GRASP can perform both <a href="#technical-info">joint</a> and <a href="#technical-info">marginal</a> reconstructions. GRASP will initially run a joint reconstruction (see <a href="#home-img">(6)</a> in <a href="#home-img">Figure 1</a>);
                        we have made marginal reconstruction available to be run on specific nodes in the results section of GRASP.</p>
                    <h3 id="perform-recon-nav">Perform (and save) reconstruction</h3>
                    <p>Once test data has been selected or user data provided, the reconstruction can be started (see <a href="#home-img">(7)</a> in <a href="#home-img">Figure 1</a>).</p>
                    <p>We strongly recommend that you save your reconstruction before running it.
                        This means that you have to be signed in as a user, and that immediately after starting the reconstruction you will be forwarded directly to your collection of saved reconstructions.
                       You will need to wait for an email to notify you if the reconstruction was successful. If successful, you will be able to refresh you list and select it from there. If you received an error, please see below for possible causes and remedies.
                    </p>
                    <p>If you opted not to save your reconstruction, the screen will be blocked and you will need to wait until the reconstruction finishes.
                        Any errors will be displayed. If the reconstruction is successful, you will be forwarded to the results page (see below).
                        Many functions will not be available until you login and subsequently save the reconstruction. </p>
                    <p>Any saved reconstructions, can be re-loaded without completely re-running it. You can also share your reconstruction with other users--as long as you have their user names.</p>

                </div>

                <a id="home-img"></a>
                <div class="col-xs-12">
                    <div class="img" id="home-img-div"></div>
                </div>
                <figcaption class="figure-caption">Figure 1. The GRASP landing page with a form for (1) trying an example data set, (2) naming your data set, (3) selecting what aligned sequences to upload,
                    (4) selecting what phylogenetic tree to upload, (5) selecting evolutionary model, (6) opting to save or not, and (7) starting the analysis. </figcaption>

                <h3 id="errors-nav">Errors</h3>
                <p>The following is a list of error messages that can be generated by GRASP and suggestions as to how to proceed if you encounter them.</p>

                <table class="table table-hover">
                    <thead>
                    <tr>
                        <th>Error message</th>
                        <th>Suggested actions</th>
                    </tr>
                    </thead>
                    <tbody>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Incorrect sequence or alignment format (requires FASTA or Clustal format .aln, .fa or .fasta)</strong></li>
                            </ul></td>

                        <td>Your input files are in the incorrect format. Check that they are valid FASTA or Clustal
                            formats with the correct file extensions. Check that your Clustal file has a correct
                            header (starting with the word “CLUSTAL”). If they are in a different format there are
                            tools available online to convert them to FASTA or Clustal
                            (<a href="https://www.ebi.ac.uk/Tools/sfc/emboss_seqret/">https://www.ebi.ac.uk/Tools/sfc/emboss_seqret/</a>).
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Aligned sequences must have the same length</strong></li>
                            </ul></td>

                        <td>An alignment file by definition should contain sequences with the same length. Check
                            that your alignment file contains aligned sequences and that they are all the same
                            length..
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>There was a problem reading the Newick file.</strong></li>
                            </ul></td>

                        <td>Check your Newick file, specifically it is worth checking that you have a single Newick
                            string that contains the correct number of opening and closing brackets, and that you do
                            not have duplicate identifiers.
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>A distance value in your Newick file couldn’t be parsed as a number.</strong></li>
                            </ul></td>

                        <td>GRASP splits Newick files on the “:” symbol and expects a distance to come immediately
                            after each “:” symbol in the file. Check that the indicated value only contains numbers.
                            If your identifiers themselves contain the “:” symbol they should be replaced (taking
                            care to replace them in the corresponding sequence alignment).
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Extant node names must be unique - &lt;...&gt; is duplicated.</strong></li>
                                <li><strong>Sequence names must be unique - &lt;...&gt; - is duplicated.</strong></li>
                                <li><strong>The sequence names in the provided alignment must all have a match.</strong></li>
                            </ul></td>


                        <td style="vertical-align:middle">Either your files contain sequence names that are
                            duplicated or sequence names do not match the names of nodes in the phylogenetic tree.
                            You will be notified of names that are unique for each file. GRASP will not guess what
                            sequence that should be placed in the tree, so make sure the names agree in the two
                            files. Some alignment programs willnamend the names, which means you may need to edit
                            them manually.
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Sequence &lt;...&gt; is using an invalid symbol. Unrecognised symbol: &lt;...&gt; at index &lt;...&gt;</strong></li>
                            </ul></td>

                        <td>Check your sequence alignment and phylogenetic tree files and replace or remove the
                            invalid symbols. In cases where ‘X’ has been used in place of an unknown amino acid
                            these can either be corrected by replacing the ‘X’ symbols with a ‘-‘ symbol or by
                            identifying higher quality sequences without unknown amino acids and realigning and
                            inferring new phylogenetic trees.
                        </td>
                    </tr>
                    <tr>
                        <td class="col-md-6">
                            <ul>
                                <li><strong>Your files are too large for the web server.</strong></li>
                            </ul></td>

                        <td>Download and install a local, commandline tool that you can run on your
                            own hardware. This will not have the graphical user interface, however.
                            See <a href="https://bodenlab.github.io/GRASP-suite">GRASP-suite</a>.
                        </td>
                    </tr>
                    </tbody>
                </table>

                <p>You should also be aware that the server only keeps the results for a limited time. You may notice that if you leave the results page open for a long time without interaction, it may stop responding.</p>

                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell"><div class="text_cell_render border-box-sizing rendered_html"><hr></hr></div></div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h2 id="results-nav">GRASP outputs</h2>
                            <h3 id="results-phylo-nav">Phylogenetic tree</h3>
                            <p><a id='phylogenetic_tree'></a></p>
                            <p>The results page shows an interactive version of the phylogenetic tree that was provided for reconstruction.
                                By default, labels of extant sequences are displayed and the user is given the option to also display or hide labels and branch lengths (see <a href="#phylo-tree-img">(2)</a> in <a href="#phylo-tree-img">Figure 2</a>).
                                It is possible to search for nodes given search pattern (see <a href="#phylo-tree-img">(3)</a> in <a href="#phylo-tree-img">Figure 2</a>).
                                If extant labels follow a standard format, it is possible to download taxonomic information from NCBI and UniProt by the press of a button (see <a href="#phylo-tree-img">(2)</a> in <a href="#phylo-tree-img">Figure 2</a>).
                            </p>
                            <p>GRASP will number ancestors (i.e. phylogenetic branch points) internally from the root (N0, N1, ...) using the principle of 'depth first'. N0 is always going to be the root of the tree.
                                You can see an ancestor label by hovering over the node, or in a pop-up box by clicking it.
                                If available, taxonomic information will be collated and summaries for ancestor nodes are shown when clicked at (more information below).
                            </p>
                            <p>It is possible to upload annotations for extant labels (see <a href="#phylo-tree-img">(1)</a> in <a href="#phylo-tree-img">Figure 2</a>)
                                that can subsequently be searched for (more information below).
                            </p>
                            <p>It is possible to search for sequence motifs in the ancestor sequences (see <a href="#phylo-tree-img">(3)</a> in <a href="#phylo-tree-img">Figure 2</a>); the syntax is described below.
                            </p>
                            <p>For big trees you may want to reduce the number of nodes shown. Use the buttons at (4) to adjust how many nodes are collapsed and the size of text labels.</p>
                            <p>Each node is coloured by evolutionary distance according to the legend on the left hand side (see <a href="#phylo-tree-img">(6)</a> in <a href="#phylo-tree-img">Figure 2</a>).
                                One node will always appear pink and be noted in the header above the tree.
                                This is the node that is having its reconstruction displayed in the graph below the phylogenetic tree and is known as the <strong>primary node</strong>.
                                By default, N0 (the root node) is the <em>primary node</em>, will appear pink and be listed in the header.</p>
                            <a id="phylo-tree-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="phylo-tree-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 2. The phylogenetic tree with options to
                                (1) associate annotations with extant nodes,
                                (2) toggle what is shown in the tree,
                                (3) search options,
                                (4) adjust the size of the tree and how much of it is displayed, and
                                (5) search and perform functions for a specified node.
                                This particular tree is contained in the 'Afriat-Jurnou et al. (29)' example available on GRASP.
                            </figcaption>
                            <p>By hovering over nodes in the phylogenetic tree, the name will be revealed. By right clicking an ancestral node in the tree you will bring up a number of
                                operations which can be performed on the ancestor. Note that some operations may take time to complete.
                                (Unless previously run for the specified ancestor, marginal reconstruction will usually take as long as the initial reconstruction.)</p>
                            <table class="table">
                                <thead>
                                <tr>
                                    <th>Option</th>
                                    <th>Operation</th>
                                </tr>
                                </thead>
                                <tbody>
                                <tr>
                                    <td>View marginal reconstruction</td>
                                    <td>Change visualisation to display the marginal reconstruction of this ancestor;
                                        make this ancestor the <em>primary node</em> in the phylogenetic tree.
                                        The ancestor in the phylogenetic tree will be highlighted in pink.</td>
                                </tr>
                                <tr>
                                    <td>View joint reconstruction</td>
                                    <td>Change visualisation to display the joint reconstruction of this ancestor;
                                        make this ancestor the <em>primary node</em> in the phylogenetic tree.
                                        The ancestor in the phylogenetic tree will be highlighted in pink.</td>
                                </tr>
                                <tr>
                                    <td>Add joint reconstruction</td>
                                    <td>Add a panel to the visualisation of another ancestor, based on the current joint reconstruction.
                                        Additional graphs are "stacked" and shown with a blue background (the hue is determined by the evolutionary distance to the root).
                                        </td>
                                </tr>
                                <tr>
                                    <td>Collapse subtree</td>
                                    <td>Hide all descendants/children of this ancestor. Use this when your tree is large, and hard to read.</td>
                                </tr>
                                <tr>
                                    <td>Expand subtree</td>
                                    <td>Expand the view of descendants/children of this ancestor. Note that this operation needs to be repeated for big subtrees (to avoid flooding the screen).</td>
                                </tr>
                                </tbody>
                            </table>
                            <div class="col-xs-8 col-xs-offset-2">
                                <div class="col-xs-12">
                                    <div class="img" id="phylo-tree-right-click-img-div"></div>
                                </div>
                                <figcaption class="figure-caption">Figure 3. The pop-up menu of the phylogenetic tree is activated by right-clicking on an internal node. </figcaption>
                            </div>
                            <div class="clearfix"></div>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h4 id="user-annotation-nav">Annotation: User-provided labels</h4>
                            <p>GRASP can assign tabulated series of values to both extants and ancestors by their labels (see <a href="#phylo-tree-img">(1)</a> in <a href="#phylo-tree-img">Figure 2</a>).
                                For example, a user can use this to assign functional annotation indexed by a column header, say 'Substrate', to a certain protein, say 'N123' or 'P12345'.
                                This would be uploaded in the form of a tab-separated value (TSV) file, e.g.</p>
                            <pre><code>
        ID          Name         Substrate    Activity
        N123        Fav_ancestor Luciferase   High(100)
        P12345      My-enzyme    Luciferase   Moderate(10)
        P54321      No-enzyme    None         None(0)
                            </code></pre>
                            <p>Upload this file, and use the same popup menu to enter which column you want to search in, say 'Substrate', and the term that needs to be matched for a node to be highlighted, say 'Luciferase'.
                            </p>

                            <h4 id="taxonomy-annotation-nav">Tree vis options: Taxonomy</h4>
                            <p> GRASP can automatically retrieve taxonomic information from the NCBI and UniProt databases by clicking on the <em>Annotate taxonomy</em> button, and append this information to the extant nodes of the phylogenetic tree.
                                Ancestor nodes in the tree are assigned the highest shared taxonomic rank of the sequences beneath them.
                                Once nodes are annotated, the highest shared rank will be visible at <em>collapsed</em> nodes.
                                Clicking on an ancestor node will report the highest shared taxonomic rank; the info box will present a histogram of the distribution within the
                                taxonomic rank immediately below the highest shared rank</p>
                            <p>For example, in Figure 4, the four sequences contained under the ancestral node N1 all belong the the Phylum Ascomycota, but differ in their taxonomic Class. </p>
                            <div class="col-xs-8 col-xs-offset-2">
                                <div class="col-xs-12">
                                    <div class="img" id="taxonomic-annotation-img-div"></div>
                                </div>
                                <figcaption class="figure-caption">Figure 4. The summary available after annotating nodes with their taxonomic rank. </figcaption>
                            </div>
                            <div class="clearfix"></div>
                            <p>In order to annotate taxonomy from the NCBI and UniProt databases, GRASP expects the sequence IDs in your FASTA file to be in either the NCBI or UniProt ID format -</p>
                            <p><code>>PRD21445.1 Cyp2u1 [Nephila clavipes]</code></p>
                            <p>or</p>
                            <p><code>>tr|A0A2P6K539|A0A2P6K539_NEPCL Cyp2u1 OS=Nephila clavipes OX=6915 GN=NCL1_51699 PE=4 SV=1</code></p>
                            <p>By default, GRASP will expect the ID to be separated by a whitespace from any additional information.</p>
                            <p>However, it will accept IDs which are completely stripped of additional information - </p>
                            <p><code>>PRD21445.1</code></p>
                            <p>or</p>
                            <p><code>>tr|A0A2P6K539|A0A2P6K539_NEPCL</code></p>
                            <p>Or IDs that have all whitespace removed (in order to force annotations to appear on phylogenetic trees) <em>if</em> the additional information is separated from the ID by a pipe symbol (|) -</p>
                            <p><code>>PRD21445.1|Cyp2u1_Nephila_clavipes</code></p>
                            <p>or</p>
                            <p><code>>tr|A0A2P6K539|A0A2P6K539_NEPCL|Cyp2u1_OS=Nephila_clavipes</code></p>
                            <p>Note that this means that headers where additional information is not separated by either a whitespace or a pipe symbol will not be able to have their taxonomy annotated - i.e. the following will <em>not</em> work -</p>
                            <p><code>>PRD21445.1_Cyp2u1_Nephila_clavipes</code></p>
                            <p>or</p>
                            <p><code>>tr|A0A2P6K539|A0A2P6K539_NEPCL_Cyp2u1_OS=Nephila_clavipes</code></p>
                            <p>All of the existing constraints about sequence names matching between alignment and phylogenetic tree must still be adhered to.</p>

                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h4 id="motif-searching-nav">Tree search options: Sequence motif</h4>
                            <p> You are able to search for ancestors with specified sequence motifs in saved joint reconstructions; note that searching is done only in <em>preferred</em> sequences, and not in optional paths (see more below). </p>
                            <p><h5>Example search patterns:</h5>
                            <ul>
                                <li>
                                <code> %A%</code>: Highlight sequences with an <code> A </code> in any position, you can also use just an <code> A </code> however the <code>%</code> signs are recommended.
                                </li>
                            <li>
                                <code> %A-*M%</code>: Highlight sequences with an <code> A </code> followed by any number of gaps and then an <code> M </code>.
                            </li>
                            <li>
                                <code> %AM(P|R)%</code>:  Highlight sequences with an <code> A </code> followed by an <code> M </code>, followed by either a <code> P </code> or an <code> R </code>.
                            </li>
                            <li>
                                <code> %A_M%</code>:  Highlight sequences with an <code> A </code> followed by any other character (e.g. a gap, <code> K </code>, <code> P </code> etc) followed by an <code> M </code>.
                            </li>
                            </ul>
                            </p>
                            <p><h5>Syntax guide:</h5>
                            <ul>
                                <li>
                                    <code> _ </code>: Wildcard character.
                                </li>
                                <li>
                                    <code> | </code>:  denotes one of the two options.
                                </li>
                                <li>
                                    <code> *</code>: repetition of preceding amino acid 0 or more times.
                                </li>
                                <li>
                                    <code> {n}</code>: repetition of the preceding amino acid <code>n</code> times.
                                </li>
                                <li>
                                    <code> {m, n}</code>: repetition of the preceding amino acid more than <code>m</code> times and less than <code>n</code> times.
                                </li>
                            </ul>
                            </p>
                            <p><em>Note 1:</em> Since the gap patterns from the original alignment are preserved, unless you specify gaps or use (<code>-*</code>) between your search elements, only those sequences without gaps between the search terms will be highlighted.
                            </p>
                            <p><em>Note 2:</em> The searching looks for a match <em>anywhere</em> in the sequence, this means that even though two sequences are highlighted, they might not have the motif in the same position! So it is recommended you confirm your findings by downloading the joint reconstructions and checking the alignment.
                            </p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="po-graph-nav">Partial-order graph</h3>
                            <p>GRASP uses partial-order graphs to represent ancestors.
                                They are suited to this task as (a) sequence homology amongst members in a protein family (and by extension their ancestors) may not always map unambiguously to a single, linear ordering, and
                                (b) unresolved insertions and deletions can be identified as alternative paths through the graph.</p>
                            <p>We suggest that you read a partial-order graph from left-to-right, node-by-node, forming a putative biological sequence (N- to C-terminus) by following a single path.
                                While GRASP attempts to resolve the history of insertions and deletions, at a single node, you may be presented with multiple in- and out-going edges, each of which joins potential ancestral sequences together.</p>

                            <h4>Nodes in a partial-order graph</h4>
                            <p>A node represents the character state at a position in a sequence. The position is not defined in an absolute sense, rather in a relative sense when there is evidence of an order.
                                For practical reasons, there is still a positional index to assist navigation in GRASP.</p>
                            <p>The node displays either a single character state or a distribution over possible character states. The states of ancestral nodes are inferred by <a href="/ml">maximum likelihood</a>.</p>

                            <h4>Edges in a partial-order graph and preferred paths</h4>
                            <p><a id='edge_cons_paths'></a></p>
                            <p>An edge represents a valid path to create a full-length sequence. Once homologous sections have been determined by alignment,
                                the joint set of extant sequences express the space of possible paths that can be taken. Of them, by enumerating them at each node,
                                GRASP infers the maximally <em>parsimonious</em> edges, i.e. the edge or edges that imply the least number of changes across the whole evolutionary history.
                                We call this process <em>bi-directional edge-parsimony</em> (BDEP). The term "bi-directional" here refers to the separation of in- and out-going edges of a node,
                                and therefore an edge can be maximally parsimonious, for one or both nodes it joins. You can find a more detailed explanation <a href="#technical-info">here</a>.</p>
                            <p>GRASP identifies edges for which there is qualified support, and makes distinctions that are important to interpret sequences that you assemble through them:</p>
                            <ul>
                                <li>
                                    <p>Solid edges are fully supported by bi-directional edge-parsimony; these links are strong candidates for how an ancestor can be assembled.
                                        They are not exclusive; multiple paths connecting different parts of a full-length sequence may be present.
                                        Nor are they guaranteed to identify even a single full-length sequence.</p>
                                </li>
                                <li>
                                    <p>A dotted edge is maximally parsimonious when only one of the nodes it joins is considered (which one is not shown).
                                        These edges are weaker candidates but are kept to allow full-length ancestors to be formulated in the absence of the stronger option above.</p>
                                </li>
                                <div class="col-xs-12">
                                    <div class="img" id="dark-light-edges-img-div"></div>
                                </div>
                                <div class="clearfix"></div>
                            </ul>
                            <p>GRASP also uses the number of extant sequences on which the inference is based. This can be used to separate edges further:</p>
                            <ul>
                                <li>
                                    <p>The darker the edge, the more extant sequences (of the set in the subtree under the ancestor) travel through the same edge.</p>
                                </li>
                                <li>
                                    <p>GRASP provides the option of displaying one <a href="#pref-path-nav"><em>preferred</em> path</a> through a sequence. If this option is turned on,
                                        thick/bold edges represent a single, preferred traversal of the graph, giving preference to
                                        (a) bi-directional edge-parsimony support (edge is drawn dark) and (b) greater number of extant sequences following (in the subtree under the ancestor).
                                        </p>
                                </li>
                            </ul>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="vis-joint-nav">Graph visualisation - joint reconstruction</h3>
                            <p>After running reconstruction, you will be able to see a navigation bar (see <a href="#graph-view-img">(1)</a> in <a href="#graph-view-img">Figure 5</a>),
                                a multiple sequence alignment graph (see <a href="#graph-view-img">(2)</a> in <a href="#graph-view-img">Figure 5</a>) and
                                the <em>primary</em> ancestor graph(see <a href="#graph-view-img">(3)</a> in <a href="#graph-view-img">Figure 5</a>) underneath the phylogenetic tree.</p>
                            <p>GRASP allows you to 'add' ancestors to be visualised along side the primary ancestor.
                                Each added ancestor becomes the <em>current</em> ancestor graph.</p>
                            <a id="graph-view-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="graph-view-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 5. The reconstruction results panel has multiple sub-panels to view and interact with, including
                                (1) a bar for viewing the overall variation (optional edges and insertion/deletions) found in the set of (displayed) reconstructions,
                                (2) a partial-order graph representing the input sequence alignment,
                                (3) a partial-order graph for the <em>primary</em> ancestor,
                                (4) slider for selecting a section, and zooming in and out of the set of ancestors currently displayed, and a set of buttons (5);
                                the buttons allow the user to save the graphs graphically, select colours for amino acids as they appear in graphs, perform mutant library analysis (applies to marginal reconstructions only), and view only the preferred path.
                            </figcaption>
                            <h4>Navigation bar</h4>
                            <p>The navigation bar has a blue box (see <a href="#graph-view-img">(4)</a> in <a href="#graph-view-img">Figure 5</a>) which can be expanded, or collapsed and moved across the window to view different sections of the graphs.
                                The region of the graph that you are viewing is within the blue box. Relative to the primary ancestor that is being viewed (see <a href="#phylogenetic_tree">Phylogenetic tree</a>) you will see grey boxes and red circles across the navigation bar.</p>
                            <p><em>Grey boxes</em> represent insertions and deletions in the ancestor that cannot be resolved definitively; they indicate the presence of two or more edges in the ancestor at the corresponding position.</p>
                            <p><em>Red circles</em> represent definitive absence of content in the current ancestor, relative to the input sequence alignment graph.</p>
                            <p>If you add multiple ancestor POGs shown together, grey boxes are determined from the union of ancestors, but red circles apply only to the <em>current</em> ancestor.</p>
                            <h4>Input sequences</h4>
                            <p>The multiple sequence alignment is represented here as a partial-order graph that contains paths to regenerate all sequences uploaded by the user.
                                Each node in the graph is depicted as a pie chart representing the distribution of amino acids at that position across all extant sequences.
                                The most probable amino acid is shown using its one-letter code.
                                When you hover your mouse over each node, a bar chart will display, showing the distribution in more detail, as seen in <a href="#msa-pie-chart-img">Figure 6</a>.</p>
                            <a id="msa-pie-chart-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="msa-pie-chart-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 6. The input sequence alignment graph shows the sequence composition of <i>all</i> extant sequences, and what insertion/deletions are implied, in terms of alternative paths of edges.</figcaption>
                            <p>The edges in the multiple sequence alignment graph represent the insertion and deletions of <em>all</em> extant sequences.</p>
                            <h4>Primary ancestor</h4>
                            <p>This graph displays the partial-order graph of the primary ancestor. More technical information of joint reconstruction is provided <a href="#technical-info">below</a>.</p>
                            <p>Edges in reconstructions are inferred by edge-parsimony. One or more ancestral sequences can be formed by following one path. We recommend that solid edges are given preference, but all presented edges have some support.</p>
                            <h4>The joint reconstruction 'stack'</h4>
                            <p>When the 'Add joint reconstruction' option is selected for a node in the phylogenetic tree, its reconstructed sequence will be added below the primary ancestor (boxed with a blue shade that indicates the evolutionary distance of the phylogenetic node it is taken from to the root).
                                We refer to this as the 'stack' and you can keep adding joint reconstructions to it; this allows for comparisons between ancestral states at different phylogenetic branch points.</p>
                            <a id="merged-graph-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="merged-graph-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 7. After joint reconstruction, the primary ancestor and added ancestors are shown as a stack of graphs. The user can repeatedly 'add' ancestors by selecting them from the phylogenetic tree (the tree is shown in blue-box inset). </figcaption>
                            <h4>Graph options</h4>
                            <p>Graphs can be saved as PNG images using the 'Save graphs' button (see <a href="#graph-view-img">(5)</a> in <a href="#graph-view-img">Figure 5</a>).
                                The amino acid colour scheme is, by default, set to Clustal.</p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="vis-marg-nav">Graph visualisation - marginal reconstruction</h3>
                            <p>A marginal reconstruction can be performed on any ancestor selected in the
                                phylogenetic tree on the results page.
                                A marginal reconstruction will look similar to <a href="#graph-view-marginal-img">Figure 8</a>.
                            </p>
                            <p>The reconstruction of the primary node will be composed of pie charts similar to the input sequence alignment graph.
                                Each pie chart depicts the distribution of amino acids at each position in the reconstructed ancestor.
                                More technical information about marginal reconstruction is provided <a href="#technical-info">here</a>.
                            </p>
                            <a id="graph-view-marginal-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="graph-view-marginal-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 8. After marginal reconstruction, the primary ancestor is shown as a graph, where each node is a pie chart displaying the probability of each amino acid at that position. Distributions can be inspected by holding the mouse pointer over the node.
                                The ability to suggest and view mutants around an ancestor is enabled ('View mutants' in <a href="#graph-view-img">(5)</a> in <a href="#graph-view-img">Figure 5</a>).
                            </figcaption>
                            <h4>Mutants</h4>
                            <p>Once a marginal reconstruction has been performed, it is possible to identify targets for mutational analysis.
                                This is based on the idea that the reconstruction represents just one possible ancestor, and that others can be explored around it by mutating positions that are inferred to be in an ambivalent state.</p>
                            <p>The user can select the level of complexity of the mutant library.
                                The base case "0" means that all nodes map to the single, most probable amino acid.
                                You will note by setting it to "0" all nodes become single-state.
                                By increasing the level, alternative amino acids are added to positions that are <em>more</em> ambivalent;
                                at higher levels of complexity, some positions are assigned two or more mutational targets.
                                An arrow appears close to the location of the identified nodes alongside the navigation bar.
                                You will note that the ambivalent nodes are shown with more character states.
                                More technical information is provided <a href="#technical-info">below</a>.
                            </p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h3 id="download-nav">Downloading results</h3>
                            <p>You can download results generated by GRASP, through the download section.
                                The download section is expanded by clicking the "Download Results" drop down button located below the graph visualisation area (see <a href="#download-pane-img">(1)</a> in <a href="#download-pane-img">Figure 9</a>).</p>
                            <p>Note that sequences are based on a single, <a href="#pref-path-nav"><em>preferred</em> path</a> in each, more information-rich ancestor partial-order graph.
                            We recommend that you still consult with the partial-order graph to corroborate what the evidence for the ancestor sequence is.</p>
                            <a id="download-pane-img"></a>
                            <div class="col-xs-12">
                                <div class="img" id="download-pane-img-div"></div>
                            </div>
                            <figcaption class="figure-caption">Figure 9. The download section is expanded by clicking (1), enabling the download of (2) the annotated phylogenetic tree, (3) marginal reconstructions and (4) joint reconstructions.
                                </figcaption>
                            <p>To download results, click the appropriate button.</p>
                            <p>The tree <a href="#download-pane-img">(2)</a> (in <a href="#download-pane-img">Figure 9</a>) is the tree the user provided but with internal/ancestral nodes labelled (N0, N1..etc.).</p>
                            <p>A marginal reconstruction <a href="#download-pane-img">(3)</a> (in <a href="#download-pane-img">Figure 9</a>) is available as either
                                TSV file (tabulated for the probability of each character, in each position) and
                                FASTA file (with the most probable character in each position).
                                Note that only the single, <em>primary</em> ancestor is downloaded.</p>
                            <p>A joint reconstruction <a href="#download-pane-img">(4)</a> (in <a href="#download-pane-img">Figure 9</a>) is available as a FASTA file (with nominated ancestors, or the full set of ancestors).</p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h2 id="tech-info-nav">Technical information<a id="technical-info"></a></h2>
                            <h3 id="joint-marg-nav">What is the difference between joint and marginal reconstruction?</h3>
                            <p>GRASP employs exact probabilistic inference; it distinguishes between two types of reconstructions:</p>
                            <p><strong>Joint</strong> - GRASP infers the most probable combination of all ancestral states.
                                The assignment of all ancestral states (across the whole phylogenetic tree) optimises a single objective
                                (the joint state with the greatest <em><a href="/ml">likelihood</a></em>).
                                This will tell you which are the most probable ancestors, given the extant sequences and their phylogenetic relationships.
                                The inference is valid for the whole tree, and it is practical to compare ancestors with one another as they represent
                                snapshots from the same evolutionary history.</p>
                            <p><strong>Marginal</strong> - GRASP infers the probability distribution of each residue for a single, nominated ancestor.
                                This inference will present you with the probability of amino acids at each position in the nominated ancestor,
                                given extant sequences and their phylogenetic relationships. The inference "marginalises" all other ancestors,
                                which means that (a) each ancestor needs to be inferred separately, and (b) resulting distributions are not directly comparable.
                                Multiple marginal reconstructions (for different ancestors) are based on the <em>same</em> input data, but their mutual consistency is not enforced;
                                they allow for <em>different</em> versions of evolutionary histories along the <em>same</em> branches.</p>
                            <p>The most probable character state in each character distribution resulting from a marginal reconstruction
                                is commonly identical to that from a joint reconstruction; so, in practise, it may not matter much which inference type you use.</p>
                            <p>Once GRASP performs a joint reconstruction it will determine and keep all ancestral states in memory; this process can take some time to complete.
                                "Adding" nodes on the results page will not require additional inference of ancestral states, so the delay you experience is minor.
                                Marginal reconstruction needs to be performed for each node; inference is separate so there will be longer delays each time
                                even after you have moved to the results page. (That said, we now have implemented caching of these states, so if you have "added" them before, the delay is minor.)</p>

                            <h3 id="bi-pars-nav">What is bi-directional edge-parsimony?</h3>
                            <p>Edges represent optional paths through which sequences can be composed;
                                ultimately, the collective of extant sequences when aligned determines the set of all possible paths;
                                that set is the source from which we draw the path or paths for any ancestor.
                                We are interested in identifying which edges that are best supported for an ancestor, given their phylogenetic relationships.
                                For this we use the basic principle of maximum parsimony, where only the most parsimonious edges are kept for each individual ancestor.</p>
                            <p>Maximum parsimony is applied to edges of a node; for each node, we separately process edges labelled by the nodes preceding it, and edges labelled by the nodes following it.
                                An edge can therefore be supported in one direction but not the other.
                                Uni-directional support is sufficient for GRASP to display an edge and maintain any paths to the nodes concerned;
                                bi-directional support should be seen as a much stronger endorsement though.
                                We run maximum parsimony to identify <em>all</em> edges that achieve the best score (the same score can be shared by multiple edge selections).
                                This means that in many cases multiple paths through an ancestor cannot be ruled out.</p>

                            <h3 id="pref-path-nav">How is a single, preferred ancestor sequence determined from a graph?</h3>
                            <p>Not uncommonly, multiple indel histories are equally parsimonious, implying that several ancestor candidate sequences can be identified by traversing an ancestor POG;
                                however, in some applications it is necessary to nominate a single sequence.
                                To determine a 'preferred' path through an ancestor POG,
                                we determine the proportion of extant sequences in the subtree under the ancestor that follow each edge.
                                We use this proportion to express preference between multiple edges.
                            </p>
                            <p>
                                GRASP uses the A* algorithm to determine the selection of edges in a POG that jointly minimise the cost, travelling from the N- to the C-terminus.
                                GRASP imposes an absolute preference for bi-directionally parsimonious edges;
                                a uni-directional is only chosen in the absence of bi-directional edges to complete the traversal.
                                The exception is the edge to the first node, and the edge from the last node, where bi-directionality is disregarded.
                            </p>

                            <h3 id="mutant-nav">How are mutant libraries determined?</h3>
                            <p>A marginal reconstruction allows the user to identify the positions that should be mutated with priority, to explore the space of probable ancestors.</p>
                            <p>Initially, the complexity of the mutant library is "0", suggesting for each position, only the most probable amino acid.
                                At "1", a single mutation is allowed, at the position that best improves the summed <a href="https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence">Kullback-Leibler divergence</a> between the
                                mutant distributions and the corresponding marginal distributions, across all sequence positions.
                                For each increase in the level, another mutation is identified where an alternative amino acid is suggested.
                                Note that additional mutations could occur at a position that has been mutated previously or in a different position.
                                Either way, the potential library will increase in complexity, i.e. increase the number of theoretically possible mutants.
                            </p>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <a href="/ml"><h3 id="advanced-nav">Inference in GRASP</h3></a>
                        </div>
                    </div>
                </div>
                <div class="cell border-box-sizing text_cell rendered">
                    <div class="inner_cell">
                        <div class="text_cell_render border-box-sizing rendered_html">
                            <h2 id="refs-all">References</h2>
                            <h3 id="refs-grasp">Reference to GRASP and example data sets</h3>
                                <p>Foley G, et al. Identifying and engineering ancient variants of enzymes using Graphical Representation of Ancestral Sequence Predictions (GRASP).
                                    <em>bioRxiv</em> 2019. <a href="https://doi.org/10.1101/2019.12.30.891457">DOI:10.1101/2019.12.30.891457</a></p>
                            <h3 id="refs-nav">More example data sets</h3>
                            <ol>
                                <li>Afriat-Jurnou L, Jackson CJ and Tawfik DS. Reconstructing a Missing Link in the Evolution of a Recently Diverged Phosphotriesterase by Active-Site Loop Remodeling.
                                    <em>Biochemistry</em> 2012, 51, 6047-6055. <a href="http://dx.doi.org/10.1021/bi300694t">DOI:10.1021/bi300694t</a></li>
                                <li>Hudson WH, Kossmann BR, de Vera IM, Chuo SW, Weikum ER, Eick GN, Thornton JW, Ivanov IN, Kojetin DJ, and Ortlund EA.
                                    Distal substitutions drive divergent DNA specificity among paralogous transcription factors through subdivision of conformational space.
                                    <em>Proc Natl Acad Sci USA</em>, 113(2): 326-331. <a href="http://dx.doi.org/10.1073/pnas.1518960113">DOI:10.1073/pnas.1518960113</a></li>
                            </ol>
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <!-- /.container -->
    </div>
</div>
</body>
</html>